File: src/main/resources/statistics/OperationsActivity.py
#
# Copyright 2020 XEBIALABS
#
# Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
#

# https://docs.xebialabs.com/jython-docs/#!/xl-release/9.0.x//service/com.xebialabs.xlrelease.domain.status.TaskStatus

import com.xebialabs.xlrelease.api.v1.forms.ReleasesFilters as ReleasesFilters

releaseFilters = ReleasesFilters()
for releaseTag in releaseTags:
    releaseFilters.tags = [releaseTag]
if fromDateTime not in [None, ""]:
    releaseFilters.from = fromDateTime
releases = _releaseApi.searchReleases(releaseFilters)

connections = []
for release in releases:
    for phase in release.phases:
        for task in phase.tasks:
            for facet in task.facets:
                if upstreamAttribute == "Applications":
                    if (
                        _applicationApi.getById(facet.applicationId).title
                        in applicationNames
                    ):
                        connections.append(
                            [
                                _applicationApi.getById(facet.applicationId).title,
                                release.title,
                                _environmentApi.getById(facet.environmentId).title,
                                release.url,
                            ]
                        )
                elif upstreamAttribute == "Environments":
                    if (
                        _environmentApi.getById(facet.environmentId).title
                        in environmentNames
                    ):
                        connections.append(
                            [
                                _environmentApi.getById(facet.environmentId).title,
                                release.title,
                                _applicationApi.getById(facet.applicationId).title,
                                release.url,
                            ]
                        )

upstream_labels = []
release_labels = []
downstream_labels = []
for connection in connections:
    if connection[0] not in upstream_labels:
        upstream_labels.append(connection[0])
    if connection[1] not in release_labels:
        release_labels.append(connection[1])
    if connection[2] not in downstream_labels:
        downstream_labels.append(connection[2])

# Convert to indexes for real-time/interactive filtering
links = []
for connection in connections:
    links.append(
        [
            upstream_labels.index(connection[0]),
            release_labels.index(connection[1]),
            downstream_labels.index(connection[2]),
            connection[3],
        ]
    )

data = {
    "links": links,
    "upstream": upstream_labels,
    "upstreamAttribute": upstreamAttribute,
    "releases": release_labels,
    "downstream": downstream_labels,
    "downstreamAttribute": "Applications"
    if upstreamAttribute == "Environments"
    else "Environments",
}

File: src/main/resources/web/OperationsActivity/operations-activity.js
/*
Copyright 2020 XEBIALABS
Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:
The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.
THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
*/
function toggleView() {
    if (document.getElementById("releases-view").style.display == "none") {
        document.getElementById("releases-view").style.display = "block"
        chartItems = document.getElementsByClassName("chart-view")
        for (chartItem of chartItems) {
            chartItem.style.display = "none"
        }
    } else {
        document.getElementById("releases-view").style.display = "none"
        chartItems = document.getElementsByClassName("chart-view")
        for (chartItem of chartItems) {
            chartItem.style.display = "block"
        }
    }
}

window.addEventListener("xlrelease.load", function() {
    window.xlrelease.queryTileData(function(response) {
        var links = response.data.data.links;
        var upstream = response.data.data.upstream;
        var upstreamAttribute = response.data.data.upstreamAttribute;
        var releases = response.data.data.releases;
        var downstream = response.data.data.downstream;
        var downstreamAttribute = response.data.data.downstreamAttribute;
        var dom = document.getElementById("chart");
        var chart = echarts.init(dom);
        option = {
            parallelAxis: [{
                    dim: 0,
                    name: upstreamAttribute,
                    type: 'category',
                    data: upstream,
                    axisLabel: {
                        align: "right",
                        margin: -8
                    }
                },
                {
                    dim: 1,
                    name: 'Releases',
                    type: 'category',
                    data: releases,
                    axisLabel: {
                        align: "center",
                        backgroundColor: "rgba(256,256,256,0.84)",
                        borderRadius: 4,
                        padding: 4,
                        margin: 0
                    }
                },
                {
                    dim: 2,
                    name: downstreamAttribute,
                    type: 'category',
                    data: downstream
                }
            ],
            parallel: {
                left: '24%',
                right: '24%'
            },
            series: {
                type: 'parallel',
                lineStyle: {
                    width: 4,
                    color: "rgba(24, 128, 24, 40)"
                },
                data: links
            }
        };
        chart.on('axisareaselected', function() {
            var series = chart.getModel().getSeries()[0];
            var indices = series.getRawIndicesByActiveState('active');
            var activate = [];
            for (index of indices) {
                if (!activate.includes(links[index][1])) {
                    activate += [links[index][1]]
                }
            }
            for (link of links) {
                if (activate.includes(link[1]) || activate.length == 0) {
                    document.getElementById(link[1]).style.display = "block";
                } else {
                    document.getElementById(link[1]).style.display = "none";
                }
            }
        });
        if (option && typeof option === "object") {
            chart.setOption(option, true);
        }
        var link;
        var processed = [];
        for (link of links) {
            if (!processed.includes(releases[link[1]])) {
                processed += [releases[link[1]]]
                document.getElementById("releases").innerHTML += `<tr id="${link[1]}"><td><a href="${link[3]}" target="_blank">${releases[link[1]]}</a></td></tr>`
            }
        }
    });
});